### ISSUE TITLE
CFP: Support basic-auth/oauth2/mTLS for north-south traffic (GatewayAPI/Ingress)

### ISSUE THREAD

**Issue by chancez** [cilium/cilium#23797]
## Cilium Feature Proposal

**Describe the feature you'd like**

I'd like to be able to expose my applications with various forms of authentication. For example, I want to expose hubble-relay outside of my cluster. 

Various authentication mechanisms work better for various situations. Generally all 3 mentioned should be relatively standard APIs for configuration. One thing that would be really nice, is to be able to configure different authentication mechanisms based on different rules. For example, mTLS/client-certs if the request is coming from a certain IP range, and basic-auth/oauth2 if a specific user-agent is set.

**(Optional) Describe your proposed solution**

I would like these to be expose in GatewayAPI, but Ingress is fine too.

**Comment by artyom-p** [comment]
Planning to migrate from aws-cni + istio to cilium, we have external auth middleware on istios ingressgateway which redirects user to our keykloak instance, at the moment absence of this feature is one of the blockers for migration.

**Comment by youngnick** [comment]
@artyom-p, could you point me to any documentation about that middleware? The work to design something to do this for Gateway API is still pretty early, so the more concrete examples we have, the better.

I agree these should be exposed in Gateway API, currently Policy Attachment is the way to do this, because noone is sure of the right way to describe the config.

tl;dr Upstream Gateway API design work needs to happen before we can implement this for Gateway API. I'd prefer to wait for that than to add a whole bunch of annotations to Ingress (annotations are sticky and unstructured, which is a bad combination for supportability.)

**Comment by artyom-p** [comment]
Thats not related to Gateway API,  its a feature of Istio/Envoy itself, just a middleware. You can check example [here](https://goauthentik.io/docs/providers/proxy/forward_auth).

**Comment by chancez** [comment]
Yeah, that filter would be pretty useful. It's pretty common to avoid double proxying with an auth proxy, and using the auth proxy to act more as an authentication/authorization API. This is often used with oauth2-proxy/authelia. I agree adding support for that would be very useful for a lot of users.

**Comment by youngnick** [comment]
Ah, okay, that's using the ext_auth support, I agree that's very useful, thanks.

**Comment by laibe** [comment]
would be very much interested in this feature as well!

**Comment by toms-place** [comment]
+1 for this feature! (:

**Comment by jbgcarnes** [comment]
Very interested as well.

**Comment by rauanmayemir** [comment]
Now that cilium supports running envoy in a separate DS, it would make sense to include more built-in extensions like ext_authz, etc.

**Comment by Jamison1** [comment]
I am looking for this as well.

[Upstream GatewayAPI has an issue tracker](https://github.com/kubernetes-sigs/gateway-api/issues/1494) for auth.  Its listed as backlog due to needing help on it.

**Comment by from-nibly** [comment]
Also a blocker for us to move to cilium ingress.

**Comment by unusualfor** [comment]
This is preventing us to fully move to Cilium as well

**Comment by liyihuang** [comment]
Is it possible to have the field under network policy's HTTP for jwt authz when the gateway API does not have the API? network policy can also be used between backend services, that's useful when the backend svc using jwt to do the authz

**Comment by EternalDeiwos** [comment]
There was an excellent proposal in kubernetes-sigs/gateway-api#532 that would have allowed for authentication/authorization to be delegated to some external service like OPA. Ideally both `Ingress`, `Gateway` (global auth) and `HTTPRoute` could have something like this.

**Comment by Jamison1** [comment]
Here is an [example of GatewayAPI auth using an API key](https://gist.github.com/guicassolato/7dc98df842a89657050514d31daadaa3) as a reference as it seems the Kubernetes upstream is still listed as help needed.[](url)

**Comment by gsharma-jiggzy** [comment]
Not sure if anyone looked at the Envoy Gateway (https://gateway.envoyproxy.io/latest/user/oidc/) but it has a concept called a security policy for oidc in the next release. It would be great if the security policy would be implemented in a similar manner.

**Comment by kevin-shelaga** [comment]
+1 to this feature!

**Comment by lukapetrovic-git** [comment]
definite +1 for this!

**Comment by abctaylor** [comment]
+1

**Comment by aldycool** [comment]
+1

**Comment by svilcu** [comment]
+1

**Comment by simonostendorf** [comment]
This is needed to switch to cilium service mesh.

**Comment by ormandj** [comment]
For the mTLS side: https://gateway-api.sigs.k8s.io/geps/gep-91/

**Comment by youngnick** [comment]
Auth for north-south Gateway API connections is still being discussed, it looks likely that we'll try and push something forward in Gateway API 1.3, which will start once Gateway 1.2 is out (around Kubecon NA 2024 in mid-November).

Just wanted to let everyone know that this issue is not forgotten, we definitely want to get to it, but we need to work with upstream on doing it the right way.

**Comment by youngnick** [comment]
Oh, I should note that GEP-91 that @ormandj mentioned above is for specifying a client certificate, but it really only guarantees that the client certificate is signed by the specified CA, it can't handle any authorization based on the contents of the certificate or JWT or anything - that will require something else (which is what we're going to work on soon).

**Comment by ormandj** [comment]
@youngnick good point - this has been a major sticking point for us with moving towards Gateway API for everything; how to not just authenticate clients but handle authorization based on their identity.

As a temporary solution, you can create and manage a CA per role and generate client certificates this way, but the management overhead isn't ideal. It'd be nice if there was something available like this, at least as a temporary stop-gap until the more correct/full-featured implementation was in place.

**Comment by sholdee** [comment]
> For the mTLS side: https://gateway-api.sigs.k8s.io/geps/gep-91/

I set this up, get no errors, and everything seems to be correct, but it does not work and no client cert is requested. This does not appear to be implemented. There is another issue for it: https://github.com/cilium/cilium/issues/34359

**Comment by zviratko** [comment]
It would be great if one could add a header containing the client certificate DN (and possibly the CA chain that authenticated it). That way many applications can be extended to consume this (as long as that header can be trusted that is).

**Comment by AlexanderSarson** [comment]
+1

**Comment by youngnick** [comment]
The Gateway API Auth GEP is scheduled to be worked on as part of the v1.3 release, see kubernetes-sigs/gateway-api#1494 for the details.

Once this GEP has moved to Experimental, we will add support for it in Cilium. I can't guarantee that will happen in the Cilium 1.17 timeframe, but we will definitely add it once it's in upstream.

**Comment by andrebovendorp** [comment]
+1

**Comment by jhutchins** [comment]
Since https://github.com/kubernetes-sigs/gateway-api/issues/1494 has been accepted as experimental for version v1.4 of the Gateway API Release Cycle, as stated here https://github.com/kubernetes-sigs/gateway-api/discussions/3824, does that mean that development for this ticket is soon to be underway?

**Comment by youngnick** [comment]
We're going to iterate on GEP-1494 as part of the Gateway API v1.4 release, yes. As soon as we have agreed on something upstream and have an experimental API included in the experimental channel, Cilium will implement it.

**Comment by divStar** [comment]
Will this eventually become a "Gateway API"-only feature? I'd love Cilium ingress  to have this feature as well if somehow possible.

**Comment by Skaronator** [comment]
GEP-1494 has been merged upstream ([https://github.com/kubernetes-sigs/gateway-api/pull/4001](https://github.com/kubernetes-sigs/gateway-api/pull/4001)), and with the 1.4 code freeze in place, the release is essentially pending ([https://github.com/kubernetes-sigs/gateway-api/pull/4042](https://github.com/kubernetes-sigs/gateway-api/pull/4042)). This means it could be implemented in Cilium.

**Comment by youngnick** [comment]
Yes, that's right. This is on my list to add support for once Gateway API 1.4 releases.

**Comment by Taner314** [comment]
[Gateway API 1.4.0 is officially released](https://github.com/kubernetes-sigs/gateway-api/releases/tag/v1.4.0), and [GEP-1494](https://github.com/kubernetes-sigs/gateway-api/pull/4001) is now in experimental.

**Comment by usma0118** [comment]
@youngnick any plans around this? :)

**Comment by josemaia** [comment]
> Will this eventually become a "Gateway API"-only feature? I'd love Cilium ingress to have this feature as well if somehow possible.

This would be really nice, to have a smoother onboarding path for anyone migrating away from ingress-nginx. Gateway API has more limitations that wouldn't allow an in-place replacement.

**Comment by naivary** [comment]
> [Gateway API 1.4.0 is officially released](https://github.com/kubernetes-sigs/gateway-api/releases/tag/v1.4.0), and [GEP-1494](https://github.com/kubernetes-sigs/gateway-api/pull/4001) is now in experimental.

A really good reason to implement it after this.

**Comment by youngnick** [comment]
This is still planned, but it's a big feature, and I haven't got to it yet.

**Comment by rnxmlu** [comment]
Would be a really appreciated feature in Cilium Gateway API, otherwise it is a show-stopper while migrating away from ingress-nginx...

**Comment by gauteoh** [comment]
I created #45704 yesterday, sorry about creating a duplicate. I think however, this issue is too broad as it covers mTLS among other things. Should this issue be rephrased to only cover `HTTPRouteExternalAuth`?

**Comment by youngnick** [comment]
I think your PR should be changed to just "Updates #23797" instead of "Fixes", and that way we can explore if there are other use cases this issue should address.

**Comment by saiprateek** [comment]
I would like to work on an incremental improvement for this CFP by tightening Gateway API ExternalAuth validation.

  Cilium already translates HTTPRoute ExternalAuth to Envoy ext_authz. A useful next step is to treat `externalAuth.backendRef` like other backend references in route status validation, so missing ports, missing Services, invalid kinds, and missing
  cross-namespace ReferenceGrants surface as `ResolvedRefs=False`.

  This does not implement native basic-auth/OAuth2/mTLS inside Cilium, but it makes the supported ExternalAuth path safer and easier to operate, and gives users a reliable base for OAuth2/basic-auth auth services.

**Comment by jasonpanosso** [comment]
is this now supported in 1.20 via #45739?

**Comment by youngnick** [comment]
Yes, #45739 was included in Cilium v1.20. We haven't closed this issue out because of the tasks that @saiprateek mentioned.
